File: server/py_server/grps_framework/executor/executor.py
# Copyright 2022 netease. All rights reserved.
# Author zhaochaochao@corp.netease.com
# Date   2022/9/13
# Brief  Executor to run complete infer, including outlier, converter, dag, model infer and etc.
import time

from grps_framework.apis.grps_pb2 import GrpsMessage
from grps_framework.batching.batcher import DynamicBatcher
from grps_framework.conf.conf import global_conf
from grps_framework.context.context import GrpsContext
from grps_framework.converter.converter import converter_register
from grps_framework.dag.dag import SequentialDag
from grps_framework.logger.logger import logger
from grps_framework.model_infer.inferer import inferer_register


class Executor(object):
    def __init__(self):
        self.__outliers = {}
        self.__model_dict = {}
        self.__init_models()
        self.__dag = None
        self.__init_dag()

    def __init_models(self):
        """Init models."""
        logger.info('Init models.')
        used_customized_inferer = set()
        used_customized_converter = set()
        for model_desc in global_conf.inference_conf['models']:
            model_name = model_desc.get('name', '')
            if not model_name:
                logger.critical('Model name not specified.')
                raise ValueError('Model name not specified.')
            model_version = model_desc.get('version', '')
            if not model_version:
                logger.critical('Model version not specified.')
                raise ValueError('Model version not specified.')
            true_model_name = model_name + '-' + model_version

            if true_model_name in self.__model_dict:
                logger.critical('Model {} already exists.'.format(true_model_name))
                raise ValueError('Model {} already exists.'.format(true_model_name))

            model_path = model_desc.get('inferer_path', '')
            model_device = str(model_desc.get('device', ''))
            inp_device = str(model_desc.get('inp_device', ''))

            inferer_type = model_desc.get('inferer_type', '')
            if not inferer_type:
                logger.critical('Model inferer type not specified.')
                raise ValueError('Model inferer type not specified.')

            # Get model inferer.
            if inferer_type == 'torch':
                from grps_framework.model_infer.torch_inferer import TorchModelInferer
                model_infer = TorchModelInferer()
                inferer_name = 'torch'
            elif inferer_type == 'tensorflow':
                from grps_framework.model_infer.tf_inferer import TfModelInferer
                model_infer = TfModelInferer()
                inferer_name = 'tensorflow'
            elif inferer_type == 'tensorrt':
                from grps_framework.model_infer.trt_inferer import TrtModelInferer
                model_infer = TrtModelInferer()
                inferer_name = 'tensorrt'
            elif inferer_type == 'customized':
                inferer_name = model_desc.get('inferer_name')
                if not inferer_name:
                    logger.critical('Inferer name must be specified when model type is customized.')
                    raise ValueError('Inferer name must be specified when model type is customized.')
                if inferer_name not in inferer_register.model_inferer_dict:
                    logger.critical('Inferer name {} not found, but bound with {} model.'
                                    .format(inferer_name, true_model_name))
                    raise ValueError('Inferer name {} not found, but bound with {} model.'
                                     .format(inferer_name, true_model_name))
                if inferer_name in used_customized_inferer:
                    # Has been used, clone a new one.
                    model_infer = type(inferer_register.get_model_inferer(inferer_name))()
                else:
                    model_infer = inferer_register.get_model_inferer(inferer_name)
                    used_customized_inferer.add(inferer_name)
            else:
                # TODO: Add more model types.
                logger.critical('Inferer type {} not supported.'.format(inferer_type))
                raise ValueError('Inferer type {} not supported.'.format(inferer_type))
            model_device = model_device.lower()
            if model_device == 'original' and inferer_type == 'torch':
                model_device = model_device + '_' + inp_device
            model_device = model_device.lower()
            inferer_args = model_desc.get('inferer_args', dict())
            model_infer.init(model_path, model_device, inferer_args)
            logger.info(
                'Init model inferer {} successfully, path: {}, device: {}, args: {}'.format(
                    inferer_name, model_path, model_device, inferer_args))
            # Load model inferer.
            if not model_infer.load():
                logger.critical('Load model inferer {} failed.'.format(inferer_name))
                raise RuntimeError('Load model inferer {} failed.'.format(inferer_name))
            logger.info('Load model inferer {} successfully.'.format(inferer_name))

            converter_type = model_desc.get('converter_type', '')
            if not converter_type:
                logger.critical('Model converter type not specified.')
                raise ValueError('Model converter type not specified.')
            converter_path = model_desc.get('converter_path', '')

            # Get converter.
            if converter_type == 'torch':
                from grps_framework.converter.torch_tensor_converter import TorchTensorConverter
                converter = TorchTensorConverter()
                converter_name = 'torch'
            elif converter_type == 'tensorflow':
                from grps_framework.converter.tf_tensor_converter import TfTensorConverter
                converter = TfTensorConverter()
                converter_name = 'tensorflow'
            elif converter_type == 'tensorrt':
                from grps_framework.converter.trt_tensor_converter import TrtTensorConverter
                converter = TrtTensorConverter()
                converter_name = 'tensorrt'
            elif converter_type == 'none':
                converter = None
                converter_name = 'none'
            elif converter_type == 'customized':
                converter_name = model_desc.get('converter_name')
                if not converter_name:
                    logger.critical('Converter name must be specified when model type is customized.')
                    raise ValueError('Converter name must be specified when model type is customized.')
                if converter_name not in converter_register.converter_dict:
                    logger.critical('Converter name {} not found, but bound with {} model.'
                                    .format(converter_name, true_model_name))
                    raise ValueError('Converter name {} not found, but bound with {} model.'
                                     .format(converter_name, true_model_name))
                if converter_name in used_customized_converter:
                    converter = type(converter_register.get_converter(converter_name))()
                else:
                    converter = converter_register.get_converter(converter_name)
                    used_customized_converter.add(converter_name)
            else:
                # TODO: Add more converter types.
                logger.critical('Converter type {} not supported.'.format(converter_type))
                raise ValueError('Converter type {} not supported.'.format(converter_type))
            if converter is not None:
                converter_args = model_desc.get('converter_args', dict())
                converter.init(converter_path, converter_args)
                logger.info('Init converter {} successfully, path: {}, args: {}'.format(converter_name,
                                                                                        converter_path,
                                                                                        converter_args))

            batcher = None
            batcher_name = ''
            max_batch_size = 0
            batch_timeout_us = 0
            batching = model_desc.get('batching', '')
            if batching:
                batching_type = batching.get('type', '')
                if batching_type == 'none':
                    batcher_name = 'none'
                elif batching_type == 'dynamic':
                    batcher = DynamicBatcher()
                    batcher_name = 'dynamic'
                else:
                    logger.critical('Batching type {} not supported.'.format(batching_type))
                    raise ValueError('Batching type {} not supported.'.format(batching_type))

                max_batch_size = batching.get('max_batch_size')
                if not max_batch_size:
                    logger.critical('Max batch size not specified.')
                    raise ValueError('Max batch size not specified.')
                if type(max_batch_size) != int:
                    logger.critical('Max batch size must be int.')
                    raise ValueError('Max batch size must be int.')
                if max_batch_size <= 0:
                    logger.critical('Max batch size must be positive.')
                    raise ValueError('Max batch size must be positive.')
                batch_timeout_us = batching.get('batch_timeout_us')
                if not batch_timeout_us:
                    logger.critical('Batch timeout not specified.')
                    raise ValueError('Batch timeout not specified.')
                if type(batch_timeout_us) != int:
                    logger.critical('Batch timeout must be int.')
                    raise ValueError('Batch timeout must be int.')
                if batch_timeout_us <= 0:
                    logger.critical('Batch timeout must be positive.')
                    raise ValueError('Batch timeout must be positive.')

            if batcher is not None:
                batcher.init(true_model_name, max_batch_size, batch_timeout_us, converter, model_infer)
                batcher.start()
                logger.info('Init and start batcher {} successfully, max batch size: {}, batch timeout: {} us.'.format(
                    batcher_name, max_batch_size, batch_timeout_us))
            self.__model_dict[true_model_name] = (model_infer, converter, batcher)
            logger.info('Init model {} successfully, inferer: {}, converter: {}, batcher: {}.'.format(
                true_model_name, inferer_name, converter_name, batcher_name))

    def __init_dag(self):
        """Init dag."""
        logger.info('Init dag.')
        dag_conf = global_conf.inference_conf['dag']
        dag_type = dag_conf['type']
        if dag_type == 'sequential':
            self.__dag = SequentialDag(dag_conf['name'])
            self.__dag.build_dag(dag_conf['nodes'], self.__model_dict)
            pass
        else:
            # TODO: Add more dag types.
            raise ValueError('Dag type {} not supported.'.format(dag_type))

        logger.info('Init dag successfully.')

    def infer_with_model_name(self, inp: GrpsMessage, context: GrpsContext, model_name=None):
        if model_name not in self.__model_dict:
            raise RuntimeError('Model {} not found.'.format(model_name))

        (model_inferer, converter, batcher) = self.__model_dict[model_name]
        context.set_converter(converter)
        context.set_inferer(model_inferer)

        if batcher:
            return batcher.infer(inp, context)

        begin = time.time()
        if converter is None:
            infer_in = inp
            infer_out = model_inferer.infer(infer_in, context)
            # logger.debug('Infer_input: {}, infer_out: {}'.format(infer_input, infer_out))
            infer_time = time.time()
            if context.has_err():
                return

            logger.info(
                'Model({}), model_infer time: {:.0f} us'.format(model_name, (infer_time - begin) * 1e6))
            return infer_out
        else:
            infer_input = converter.preprocess(inp, context)
            preprocess_time = time.time()
            if context.has_err():
                return

            infer_out = model_inferer.infer(infer_input, context)
            # logger.debug('Infer_input: {}, infer_out: {}'.format(infer_input, infer_out))
            infer_time = time.time()
            if context.has_err():
                return

            out = converter.postprocess(infer_out, context)
            postprocess_time = time.time()
            if context.has_err():
                return

            logger.info(
                'Model({}), preprocess time: {:.0f} us, model_infer time: {:.0f} us, postprocess time: {:.0f} us'
                .format(model_name, (preprocess_time - begin) * 1e6, (infer_time - preprocess_time) * 1e6,
                        (postprocess_time - infer_time) * 1e6))
            return out

    def infer(self, inp: GrpsMessage, context: GrpsContext, model_name=None):
        """
        Run complete infer.
        Args:
            context: grps context
            inp: request inp.
            model_name: model name(with `name-version` format) to predict. If not define, will use default model dag
            (defined in inference.yml) to predict.

        Returns:
            Output will be returned to client.
        """
        try:
            if self.__dag is None:
                raise RuntimeError('Dag not initialized.')
            if model_name is None or model_name == '':
                out = self.__dag.infer(inp, context)
            else:
                out = self.infer_with_model_name(inp, context, model_name)
            return out
        except Exception as e:
            raise
        finally:
            if context._grpc_streaming_run:
                # Stop grpc streaming generator if grpc streaming.
                context.stop_grpc_streaming_generator()

File: server/cpp_server/src/executor/executor.cc
/*
 * Copyright 2022 netease. All rights reserved.
 * Author zhaochaochao@corp.netease.com
 * Date   2023/02/08
 * Brief  Executor to run complete infer, including outlier, converter, dag, model infer and etc.
 */

#include "executor.h"

#include <google/protobuf/text_format.h>

#include "converter/converter.h"
#include "logger/logger.h"
#include "model_infer/inferer.h"
#ifdef GRPS_TF_ENABLE
#include "converter/tf_tensor_converter.h"
#include "model_infer/tf_inferer.h"
#endif
#ifdef GRPS_TORCH_ENABLE
#include "converter/torch_tensor_converter.h"
#include "model_infer/torch_inferer.h"
#endif
#ifdef GRPS_TRT_ENABLE
#include "converter/trt_tensor_converter.h"
#include "model_infer/trt_inferer.h"
#endif
#include "batching/batcher.h"
#include "config/global_config.h"

namespace netease::grps {
void Executor::Init() {
  LOG4(INFO, "Init executor.");
  InitOutlier();
  InitModels();
  InitDag();
}

void Executor::InitOutlier() {
  // TODO: Add outlier.
}

void Executor::InitModels() {
  LOG4(INFO, "Init models.");
  const auto& inference_config = GlobalConfig::Instance().inference_config();
  if (!inference_config._is_set.models) {
    LOG4(ERROR, "No model config.");
    throw ExecutorException("No model config.");
  }

  std::unordered_set<std::string> used_customized_inferers;
  std::unordered_set<std::string> used_customized_converters;
  for (const auto& [name, model] : inference_config.models) {
    if (models_.find(name) != models_.end()) {
      LOG4(ERROR, "Model " << name << " already exists.");
      throw ExecutorException("Model " + name + " already exists.");
    }

    std::shared_ptr<ModelInferer> inferer_ptr = nullptr;
    std::string inferer_name;
    if (model.inferer_type == "torch") {
#ifdef GRPS_TORCH_ENABLE
      inferer_ptr = std::make_shared<TorchModelInferer>();
#else
      throw ExecutorException("Torch model inferer is not supported.");
#endif
      inferer_name = "torch";
    } else if (model.inferer_type == "tensorflow") {
#ifdef GRPS_TF_ENABLE
      inferer_ptr = std::make_shared<TfModelInferer>();
#else
      throw ExecutorException("Tensorflow model inferer is not supported.");
#endif
      inferer_name = "tensorflow";
    } else if (model.inferer_type == "tensorrt") {
#ifdef GRPS_TRT_ENABLE
      inferer_ptr = std::make_shared<TrtModelInferer>();
#else
      throw ExecutorException("TensorRT model inferer is not supported.");
#endif
    } else if (model.inferer_type == "customized") {
      auto customized_inferer = ModelInfererRegistry::Instance().GetInferer(model.inferer_name);
      if (customized_inferer == nullptr) {
        LOG4(ERROR, "Not found customized inferer: " << model.inferer_name);
        throw ExecutorException("Not found customized inferer: " + model.inferer_name);
      }
      inferer_name = model.inferer_name;
      if (used_customized_inferers.find(inferer_name) != used_customized_inferers.end()) {
        // If customized inferer has been used, clone new.
        inferer_ptr = std::shared_ptr<ModelInferer>(customized_inferer->Clone());
      } else {
        inferer_ptr = customized_inferer;
        used_customized_inferers.insert(inferer_name);
      }
    } else {
      LOG4(ERROR, "Not support inferer type: " << model.inferer_type);
      throw ExecutorException("Not support inferer type: " + model.inferer_type);
    }

    std::string device;
    std::transform(model.device.begin(), model.device.end(), std::back_inserter(device), ::tolower);
    if (device == "original" && model.inferer_type == "torch") {
      device += "_" + model.inp_device;
    }
    std::transform(device.begin(), device.end(), device.begin(), ::tolower);

    inferer_ptr->Init(model.inferer_path, device, model.inferer_args);
    LOG4(INFO, "Init inferer: " << inferer_name << " successfully, path: " << model.inferer_path
                                << ", device: " << device << ", args: " << model.inferer_args << ".");
    inferer_ptr->Load();
    LOG4(INFO, "Load inferer: " << inferer_name << " successfully.");

    std::shared_ptr<Converter> converter_ptr = nullptr;
    std::string converter_name;
    if (model.converter_type == "torch") {
#ifdef GRPS_TORCH_ENABLE
      converter_ptr = std::make_shared<TorchTensorConverter>();
#else
      throw ExecutorException("Torch tensor converter is not supported.");
#endif
      converter_name = "torch";
    } else if (model.converter_type == "tensorflow") {
#ifdef GRPS_TF_ENABLE
      converter_ptr = std::make_shared<TfTensorConverter>();
#else
      throw ExecutorException("Tensorflow tensor converter is not supported.");
#endif
      converter_name = "tensorflow";
    } else if (model.converter_type == "tensorrt") {
#ifdef GRPS_TRT_ENABLE
      converter_ptr = std::make_shared<TrtTensorConverter>();
#else
      throw ExecutorException("TensorRT tensor converter is not supported.");
#endif
    } else if (model.converter_type == "none") {
      converter_ptr = nullptr;
      converter_name = "none";
    } else if (model.converter_type == "customized") {
      auto customized_converter = ConverterRegistry::Instance().GetConverter(model.converter_name);
      if (customized_converter == nullptr) {
        LOG4(ERROR, "Not found customized converter: " << model.converter_name);
        throw ExecutorException("Not found customized converter: " + model.converter_name);
      }
      converter_name = model.converter_name;
      if (used_customized_converters.find(converter_name) != used_customized_converters.end()) {
        // If customized converter has been used, clone new.
        converter_ptr = std::shared_ptr<Converter>(customized_converter->Clone());
      } else {
        converter_ptr = customized_converter;
        used_customized_converters.insert(converter_name);
      }
    } else {
      LOG4(ERROR, "Not support converter type: " << model.converter_type);
      throw ExecutorException("Not support converter type: " + model.converter_type);
    }

    if (converter_ptr != nullptr) {
      converter_ptr->Init(model.converter_path, model.converter_args);
      LOG4(INFO, "Init converter: " << converter_name << " successfully, path: " << model.converter_path
                                    << ", args: " << model.converter_args << ".");
    }

    std::shared_ptr<Batcher> batcher_ptr = nullptr;
    std::string batcher_name;
    if (model.batching.type == "none") {
      batcher_name = "none";
    } else if (model.batching.type == "dynamic") {
      batcher_ptr = std::make_shared<DynamicBatcher>();
      batcher_name = "dynamic";
    } else {
      LOG4(ERROR, "Not support batching type: " << model.batching.type);
      throw ExecutorException("Not support batching type: " + model.batching.type);
    }

    if (batcher_ptr != nullptr) {
      batcher_ptr->Init(name, model.batching.max_batch_size, model.batching.batch_timeout_us, converter_ptr.get(),
                        inferer_ptr.get());
      batcher_ptr->Start();
      LOG4(INFO, "Init and start batcher: " << batcher_name
                                            << " successfully, max_batch_size: " << model.batching.max_batch_size
                                            << ", batch_timeout_us: " << model.batching.batch_timeout_us);
    }

    models_[name] = Model(model.name, model.version, converter_ptr, inferer_ptr, batcher_ptr);
    LOG4(INFO, "Init model: " << name << " successfully, inferer: " << inferer_name << ", converter: " << converter_name
                              << ", batcher: " << batcher_name << ", version: " << model.version);
  }
}

void Executor::InitDag() {
  LOG4(INFO, "Init dag.");
  const auto inference_config = GlobalConfig::Instance().inference_config();
  if (!inference_config._is_set.dag) {
    LOG4(ERROR, "No dag config.");
    throw ExecutorException("No dag config.");
  }

  const auto dag_config = inference_config.dag;
  if (dag_config.type == "sequential") {
    dag_ = std::make_shared<SequentialDag>(dag_config.name);
    dag_->BuildDag(dag_config.nodes, models_);
  } else {
    LOG4(ERROR, "Not support dag type: " << dag_config.type);
    throw ExecutorException("Not support dag type: " + dag_config.type);
  }
}

void Executor::Infer(const ::grps::protos::v1::GrpsMessage& input,
                     ::grps::protos::v1::GrpsMessage& output,
                     GrpsContext& ctx,
                     const std::string& model_name) {
#ifdef GRPS_DEBUG
  std::string input_str;
  ::google::protobuf::TextFormat::PrintToString(input, &input_str);
  LOG4(INFO, "Infer input: " << input_str);
#endif

  if (model_name.empty()) {
    dag_->Infer(input, output, ctx);
  } else {
    auto iter = models_.find(model_name);
    if (iter == models_.end()) {
      LOG4(ERROR, "Not found model: " << model_name);
      throw ExecutorException("Not found model: " + model_name);
    }

    auto& model = iter->second;
    auto converter = model.converter_;
    auto inferer = model.inferer_;
    auto batcher = model.batcher_;
    ctx.set_converter(converter.get());
    ctx.set_inferer(inferer.get());

    if (batcher) {
      batcher->Infer(input, output, ctx);
      return;
    }

    auto begin = butil::gettimeofday_us();
    if (converter) {
      std::vector<std::pair<std::string, TensorWrapper>> inp_tensors;
      std::vector<std::pair<std::string, TensorWrapper>> out_tensors;

      converter->PreProcess(input, inp_tensors, ctx);
      if (ctx.has_err()) {
        return;
      }
      auto preprocess_end = butil::gettimeofday_us();

      inferer->Infer(inp_tensors, out_tensors, ctx);
      if (ctx.has_err()) {
        return;
      }
      auto infer_end = butil::gettimeofday_us();

      output.Clear();
      converter->PostProcess(out_tensors, output, ctx);
      if (ctx.has_err()) {
        return;
      }
      auto postprocess_end = butil::gettimeofday_us();

      LOG4(INFO, "Model(" << model_name << "), preprocess latency: " << preprocess_end - begin
                          << "us, infer latency: " << infer_end - preprocess_end
                          << "us, postprocess latency: " << postprocess_end - infer_end << "us");
    } else {
      output.Clear();
      inferer->Infer(input, output, ctx);
      if (ctx.has_err()) {
        return;
      }
      auto infer_end = butil::gettimeofday_us();
      LOG4(INFO, "Model(" << model_name << "), infer latency: " << infer_end - begin << "us");
    }
  }

#ifdef GRPS_DEBUG
  std::string output_str;
  ::google::protobuf::TextFormat::PrintToString(output, &output_str);
  LOG4(INFO, "Infer output: " << output_str);
#endif
}

void Executor::Infer(const ::grps::protos::v1::GrpsMessage& input,
                     ::grps::protos::v1::GrpsMessage& output,
                     const std::shared_ptr<GrpsContext>& ctx_sp,
                     const std::string& model_name) {
#ifdef GRPS_DEBUG
  std::string input_str;
  ::google::protobuf::TextFormat::PrintToString(input, &input_str);
  LOG4(INFO, "Infer input: " << input_str);
#endif

  if (model_name.empty()) {
    dag_->Infer(input, output, ctx_sp);
  } else {
    auto iter = models_.find(model_name);
    if (iter == models_.end()) {
      LOG4(ERROR, "Not found model: " << model_name);
      throw ExecutorException("Not found model: " + model_name);
    }

    auto& model = iter->second;
    auto converter = model.converter_;
    auto inferer = model.inferer_;
    auto batcher = model.batcher_;
    ctx_sp->set_converter(converter.get());
    ctx_sp->set_inferer(inferer.get());

    if (batcher) {
      batcher->Infer(input, output, ctx_sp);
      return;
    }

    auto begin = butil::gettimeofday_us();
    if (converter) {
      std::vector<std::pair<std::string, TensorWrapper>> inp_tensors;
      std::vector<std::pair<std::string, TensorWrapper>> out_tensors;

      converter->PreProcess(input, inp_tensors, *ctx_sp);
      if (ctx_sp->has_err()) {
        return;
      }
      auto preprocess_end = butil::gettimeofday_us();

      inferer->Infer(inp_tensors, out_tensors, *ctx_sp);
      if (ctx_sp->has_err()) {
        return;
      }
      auto infer_end = butil::gettimeofday_us();

      output.Clear();
      converter->PostProcess(out_tensors, output, *ctx_sp);
      if (ctx_sp->has_err()) {
        return;
      }
      auto postprocess_end = butil::gettimeofday_us();

      LOG4(INFO, "Model(" << model_name << "), preprocess latency: " << preprocess_end - begin
                          << "us, infer latency: " << infer_end - preprocess_end
                          << "us, postprocess latency: " << postprocess_end - infer_end << "us");
    } else {
      output.Clear();
      inferer->Infer(input, output, *ctx_sp);
      if (ctx_sp->has_err()) {
        return;
      }
      auto infer_end = butil::gettimeofday_us();
      LOG4(INFO, "Model(" << model_name << "), infer latency: " << infer_end - begin << "us");
    }
  }

#ifdef GRPS_DEBUG
  std::string output_str;
  ::google::protobuf::TextFormat::PrintToString(output, &output_str);
  LOG4(INFO, "Infer output: " << output_str);
#endif
}

void Executor::Terminate() {
  for (auto& [name, model] : models_) {
    if (model.batcher_) {
      model.batcher_->Stop();
    }
  }

  dag_.reset();
  models_.clear();
}
} // namespace netease::grps
